# Importamos las librerías necesarias

In [3]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

# Cargamos el dataset limpio para poder crear un DataFrame

In [4]:
df = pd.read_csv('data_limpio.csv')
df.head(5)

Unnamed: 0                                        Descripcion  Precio  \
0           0  Spider-Man Marvel Legends Series Across The Sp...      24   
1           1  Spider-Man Marvel Legends Series Across The Sp...      22   
2           2  Fisher-Price Little People Musical Toddler Toy...       7   
3           3  CatToysOfficial - CAT Little Machines 5pcs Con...       5   
4           4  Spider-Man Marvel Legends Series Across The Sp...      22   

   Reseñas  Precio Real  Dias sin cargo                Envio Fecha de entrega  \
0       71        24.99              90  Ships to Bangladesh      Tue, Jun 20   
1       55        14.99              90  Ships to Bangladesh      Tue, Jun 20   
2     2895         9.99              90  Ships to Bangladesh      Tue, Jun 20   
3     2530        24.99              90  Ships to Bangladesh      Tue, Jun 20   
4       40        85.99              90  Ships to Bangladesh      Tue, Jun 20   

   Ventas en meses pasados  Stock  Edad              Otro  
0                      200      3     4      years and up  
1                      200      3     4      years and up  
2                      100      3     1  months - 5 years  
3                      100      3     3  months - 6 years  
4                       90      3     4      years and up

In [5]:
df.columns

Index(['Unnamed: 0', 'Descripcion', 'Precio', 'Reseñas', 'Precio Real',
       'Dias sin cargo', 'Envio', 'Fecha de entrega',
       'Ventas en meses pasados', 'Stock', 'Edad', 'Otro'],
      dtype='object')

Vemos que aparece en este DataFrame una columna que se llama 'Unnamed: 0' y otra que se llama 'Otro'. Como no nos sirven vamos a proceder a eliminarlas. Las mismas quedaron de la limpieza realizada anteriormente.

In [6]:
df.drop('Unnamed: 0', axis=1, inplace=True)

In [7]:
df.drop('Otro', axis=1, inplace=True)

Constatamos que se hayan eliminado las columnas en cuestión.

In [8]:
df.head(2)

Descripcion  Precio  Reseñas  \
0  Spider-Man Marvel Legends Series Across The Sp...      24       71   
1  Spider-Man Marvel Legends Series Across The Sp...      22       55   

   Precio Real  Dias sin cargo                Envio Fecha de entrega  \
0        24.99              90  Ships to Bangladesh      Tue, Jun 20   
1        14.99              90  Ships to Bangladesh      Tue, Jun 20   

   Ventas en meses pasados  Stock  Edad  
0                      200      3     4  
1                      200      3     4

Vamos a realizar un meta-análisis del dataset. Para elllo procederemos a crear dos listas. Una contendrá las columnas categóricas y la otra presentará las columnas numéricas.

In [30]:
categoricas = [i for i in df.columns if df[i].dtype=='object']
numericas = [i for i in df.columns if df[i].dtype=='int64' or df[i].dtype=='float64']

# Columnas numéricas

## La columna 'Precio'

Veamos cuál es el precio más alto presente en nuestro DataFrame

In [9]:
df['Precio'].max()

702

Sabemos que el precio más alto en nuestro DataFrame es de 702, veamos si refiere a más de un juguete o no, creando un DataFrame con el filtro del precio máximo.

In [10]:
df_precio_max = df[df['Precio']==702]
df_precio_max

Descripcion  Precio  Reseñas  \
10619  Infinity Studio x League of LegendsHallowed Se...     702      384   

       Precio Real  Dias sin cargo                Envio Fecha de entrega  \
10619        702.0              90  Ships to Bangladesh      Tue, Jun 20   

       Ventas en meses pasados  Stock  Edad  
10619                       90      2    15

El precio máximo refiere sólo a un producto con 384 reseñas, que no presenta descuentos, del cuál se vendieron 90 unidades en los meses pasados, quedan 2 de ellas y la edad recomendada es a partir de los 15 años.

Observemos ahora el precio mínimo presente en nuestro DataFrame.

In [11]:
df['Precio'].min()

0

Llama la atención que el precio mínimo sea igual a 0. Vamos a filtrar los datos cuyo precio sea igual a 0.

In [12]:
df[df['Precio']==0]

Descripcion  Precio  Reseñas  Precio Real  Dias sin cargo Envio  \
9203           0       0        0          0.0               0     0   

     Fecha de entrega  Ventas en meses pasados  Stock  Edad  
9203                0                        0      0     0

Esto podría deberse a un error a la hora de recolectar los datos. Como sólo se trata de una fila en DataFrame de 11403, vamos a proceder a eliminarla.

In [13]:
df.drop(index=9203, inplace=True)

Buscamos, nuevamente, cuál es el precio más bajo presente.

In [14]:
df['Precio'].min()

2

Este precio mínimo parece ser más coherente. Vamos a crear un DataFrame con este precio mínimo como filtro.

In [15]:
df_precio_min = df[df['Precio']==2]
df_precio_min

Descripcion  Precio  Reseñas  \
274                Funko Pop! MoviesBlack Adam - Cyclone       2    15461   
478    Marvel BattleworldSeries 2 Treachery at Twilig...       2     1142   
10213  Just Play The Umbrella Academy 3” Stylized Col...       2      384   
11270  Miyuki HigurashiWhen They Cry – SOTSUHanyu 1:7...       2       32   

       Precio Real  Dias sin cargo                Envio Fecha de entrega  \
274          12.99              90  Ships to Bangladesh      Tue, Jun 20   
478          19.99              90  Ships to Bangladesh      Tue, Jun 20   
10213         2.00              90  Ships to Bangladesh      Tue, Jun 20   
11270         2.00              90  Ships to Bangladesh      Thu, Jun 22   

       Ventas en meses pasados  Stock  Edad  
274                         90      8     4  
478                         90      1     0  
10213                       90      3    12  
11270                      100      1     0

In [16]:
for i in df_precio_min['Descripcion']:
    print(i)

Funko Pop! MoviesBlack Adam - Cyclone
Marvel BattleworldSeries 2 Treachery at Twilight Battle Ball
Just Play The Umbrella Academy 3” Stylized Collectible Figure- Diego, Multi-Color
Miyuki HigurashiWhen They Cry – SOTSUHanyu 1:7 Scale PVC Figure,Multicolor


Vemos que son cuatro juguetes los que tienen el precio más bajo. Cómo puede apreciarse, el rango etario va de los 0 a 12 años y el stock displonible de cada muñeco varía entre 1 unidad y 8.

Ahora vamos a crear una columna que nos indique el porcentaje de descuento aplicado sobre el producto, utilizando la columna del  'Precio Real' y la de 'Precio'.

In [17]:
df_precio_min['Porcentaje de descuento'] = round(((100*(df_precio_min['Precio Real']-df_precio_min['Precio']))/df['Precio Real']), 2)

C:\Users\usuario\AppData\Local\Temp\ipykernel_5512\1533223932.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_precio_min['Porcentaje de descuento'] = round(((100*(df_precio_min['Precio Real']-df_precio_min['Precio']))/df['Precio Real']), 2)


In [18]:
df_precio_min

Descripcion  Precio  Reseñas  \
274                Funko Pop! MoviesBlack Adam - Cyclone       2    15461   
478    Marvel BattleworldSeries 2 Treachery at Twilig...       2     1142   
10213  Just Play The Umbrella Academy 3” Stylized Col...       2      384   
11270  Miyuki HigurashiWhen They Cry – SOTSUHanyu 1:7...       2       32   

       Precio Real  Dias sin cargo                Envio Fecha de entrega  \
274          12.99              90  Ships to Bangladesh      Tue, Jun 20   
478          19.99              90  Ships to Bangladesh      Tue, Jun 20   
10213         2.00              90  Ships to Bangladesh      Tue, Jun 20   
11270         2.00              90  Ships to Bangladesh      Thu, Jun 22   

       Ventas en meses pasados  Stock  Edad  Porcentaje de descuento  
274                         90      8     4                    84.60  
478                         90      1     0                    89.99  
10213                       90      3    12                     0.00  
11270                      100      1     0                     0.00

Lo interesante es que, entre los productos con el precio más bajo, podemos encontrar porcentajes de descuento superiores al 84%. Es decir, tales productos están en una rebaja interesante, una verdadera oportunidad para un cliente potencial.

Ahora bien, el código anterior podemos aplicarlo a todo el DataFrame y analizar los descuentos del mismo.

In [36]:
df['Porcentaje de descuento'] = round(((100*(df['Precio Real']-df['Precio']))/df['Precio Real']), 2)

Vemos que se aplicó la nueva columna

In [37]:
df.head(2)

Descripcion  Precio  Reseñas  \
0  Spider-Man Marvel Legends Series Across The Sp...      24       71   
1  Spider-Man Marvel Legends Series Across The Sp...      22       55   

   Precio Real  Dias sin cargo                Envio Fecha de entrega  \
0        24.99              90  Ships to Bangladesh      Tue, Jun 20   
1        14.99              90  Ships to Bangladesh      Tue, Jun 20   

   Ventas en meses pasados  Stock  Edad  Porcentaje de descuento  
0                      200      3     4                     3.96  
1                      200      3     4                   -46.76

Las primeras dos filas de este DataFrame en la columna creada nos indican algunas pistas a tener en cuenta. Podemos ver que tenemos porcentajes positivos y negativos. A grandes rasgos encontraríamos tres categorías de porcentajes: i) aquellos productos a los que se les aplicó un descuento (porcentaje positivo mayor a 0), ii) los que aumentaron de precio (porcentaje negativo) y iii) los que permanecieron iguales (porcentaje igual a 0). Vamos a proceder a filtrar el DataFrame siguiendo los criterios anteriores.

Para facilitar las cosas vamos a crear vamos a crear una función que filtre el DataFrame.

In [52]:
#La función va a tomar tres argumentos: un dataframe, el nombre de la columna
#sobre la cual aplicar el filtro y la condición (mayor, menor o igual a cero)
def funcion_filtro(df, nombre_col, condicion):
    if condicion.lower() == 'igual a cero':
        df_filtrado = df[df[nombre_col]==0]
    elif condicion.lower() == 'mayor a cero':
        df_filtrado = df[df[nombre_col]>0]
    else:
        df_filtrado == df[df[nombre_col]<0]
    return df_filtrado

Filtramos el DataFrame para aquellas columnas cuyo descuento sea igual a 0.

In [53]:
sin_descuento = funcion_filtro(df, 'Porcentaje de descuento', 'igual a cero')

Confirmamos que se haya aplicado el filtro.

In [56]:
sin_descuento.head(2)

Descripcion  Precio  Reseñas  \
17  Gabby's Dollhouse, Deluxe Figure Gift Set with...      22      579   
18  Transformers Toys Heroic Optimus Prime Action ...      20    10204   

    Precio Real  Dias sin cargo                Envio Fecha de entrega  \
17         22.0              90  Ships to Bangladesh      Tue, Jun 20   
18         20.0              90  Ships to Bangladesh      Tue, Jun 20   

    Ventas en meses pasados  Stock  Edad  Porcentaje de descuento  
17                      100      3    11                      0.0  
18                      100      3     0                      0.0

Observamos cuántos artículos únicos no tienen descuento. Para ello vamos a utilizar la columna 'Descripción'.

In [65]:
sin_descuento['Descripcion'].nunique()

4900

Un elemento a analizar son las ventas en meses pasados que tuvieron estos artículos sin descuento.

Observamos cuántos artículos tuvieron x cantidad de ventas.

In [61]:
for i in sin_descuento['Ventas en meses pasados'].unique():
    cantidad_articulos = len(sin_descuento[sin_descuento['Ventas en meses pasados']==i])
    print(f'{cantidad_articulos} artículos tuvieron {i} ventas en los meses pasados', end='\n'*2)

360 artículos tuvieron 100 ventas en los meses pasados

173 artículos tuvieron 50 ventas en los meses pasados

6098 artículos tuvieron 90 ventas en los meses pasados

10 artículos tuvieron 200 ventas en los meses pasados



Veamos cuál fue el máximo nivel de ventas en los meses pasados de los artículos sin descuento.

In [67]:
sin_descuento['Ventas en meses pasados'].max()

200

Observemos el mínimo nivel de ventas en los meses pasados de estos mismos artículos.

In [68]:
sin_descuento['Ventas en meses pasados'].min()

50

Podemos observar que los artículos sin descuento vendieron entre 50-200 unidades en los meses pasados.